File: src/utils/gitDetector.ts
import fs from 'node:fs';
import path from 'node:path';

import simpleGit from 'simple-git';
import type { SimpleGit } from 'simple-git';

import type { GitPlatform } from '../config/config';
import { Logger } from './logging';

/**
 * Git platform detection result
 */
interface GitPlatformInfo {
  platform: GitPlatform;
  url: string;
  projectPath?: string;
  owner?: string;
  repo?: string;
}

/**
 * Remote repository information
 */
interface GitRemote {
  name: string;
  refs: {
    fetch: string;
    push: string;
  };
}

/**
 * Git repository detection utility
 * Detects whether a repository is GitLab or GitHub
 */
export class GitDetector {
  private git: SimpleGit;
  private logger: Logger;

  /**
   * Singleton instance of GitDetector
   */
  private static instance: GitDetector;

  /**
   * Get singleton instance of GitDetector
   * @param logger - Optional logger instance
   * @returns The GitDetector instance
   */
  public static getInstance(logger?: Logger): GitDetector {
    if (!GitDetector.instance) {
      GitDetector.instance = new GitDetector(logger);
    }
    return GitDetector.instance;
  }

  /**
   * Create a new GitDetector
   */
  constructor(logger?: Logger) {
    this.git = simpleGit({});
    this.logger = logger || new Logger();
  }

  /**
   * Auto-detect git remote details from the current working directory
   * @returns Detected git platform information or null if not detected
   */
  async detectGitPlatform(): Promise<GitPlatformInfo | null> {
    try {
      if (!(await this.isGitRepo())) {
        this.logger.warn('Not inside a git repository');
        return null;
      }

      const remote = await this.findPrimaryRemote();
      if (!remote) {
        return null;
      }

      return this.determineGitPlatform(remote.refs.fetch);
    } catch (error) {
      this.logError('Error detecting git platform', error);
      return null;
    }
  }

  /**
   * Check if the current directory is a git repository
   * @returns True if we're in a git repository
   */
  private async isGitRepo(): Promise<boolean> {
    try {
      await this.git.revparse(['--is-inside-work-tree']);
      return true;
    } catch (_error) {
      this.logError('Error', _error);
      return false;
    }
  }

  /**
   * Find the primary remote (origin or first available)
   * @returns The primary remote or null if none found
   */
  private async findPrimaryRemote(): Promise<GitRemote | null> {
    try {
      const remotes = await this.git.getRemotes(true);

      if (!remotes.length) {
        this.logger.warn('No git remotes found');
        return null;
      }

      // Prefer origin over others
      const origin = remotes.find((r: GitRemote) => r.name === 'origin') || remotes[0];

      if (!origin || !origin.refs || !origin.refs.fetch) {
        this.logger.warn('No valid git remote origin found');
        return null;
      }

      return origin;
    } catch (error) {
      this.logError('Error finding primary remote', error);
      return null;
    }
  }

  /**
   * Determine which git platform a URL belongs to
   * @param remoteUrl - The remote URL to analyze
   * @returns Information about the detected platform or null if not determined
   */
  private determineGitPlatform(remoteUrl: string): GitPlatformInfo | null {
    if (!remoteUrl) {
      this.logger.warn('No valid remote URL found');
      return null;
    }

    if (this.isGitHubUrl(remoteUrl)) {
      return this.extractGitHubInfo(remoteUrl);
    }

    if (this.isGitLabUrl(remoteUrl) || this.hasGitLabSpecificFiles()) {
      return this.extractGitLabInfo(remoteUrl);
    }

    this.logger.warn('Unable to determine git platform from remote URL');
    return null;
  }

  /**
   * Check if a URL is a GitHub URL
   * @param url - The URL to check
   * @returns True if the URL is a GitHub URL
   */
  private isGitHubUrl(url: string): boolean {
    return url.includes('github.com');
  }

  /**
   * Check if a URL is a GitLab URL
   * @param url - The URL to check
   * @returns True if the URL is a GitLab URL
   */
  private isGitLabUrl(url: string): boolean {
    return url.includes('gitlab.com') || url.includes('gitlab');
  }

  /**
   * Check if the repository has GitLab-specific files
   * @returns True if the repository appears to be a GitLab repository
   */
  private hasGitLabSpecificFiles(): boolean {
    return fs.existsSync(path.join(process.cwd(), '.gitlab-ci.yml'));
  }

  /**
   * Extract GitHub information from a URL
   * @param url - The GitHub remote URL
   * @returns GitHub platform information or null if parsing fails
   */
  private extractGitHubInfo(url: string): GitPlatformInfo | null {
    try {
      const { owner, repo } = this.parseGitHubUrl(url);
      return {
        platform: 'github',
        url: `https://github.com/${owner}/${repo}`,
        owner,
        repo,
      };
    } catch (error) {
      this.logError('Error extracting GitHub info', error);
      return null;
    }
  }

  /**
   * Extract GitLab information from a URL
   * @param url - The GitLab remote URL
   * @returns GitLab platform information or null if parsing fails
   */
  private extractGitLabInfo(url: string): GitPlatformInfo | null {
    try {
      const { url: gitlabUrl, projectPath } = this.parseGitLabUrl(url);
      return {
        platform: 'gitlab',
        url: gitlabUrl,
        projectPath,
      };
    } catch (error) {
      this.logError('Error extracting GitLab info', error);
      return null;
    }
  }

  /**
   * Parse a GitHub URL to extract owner and repo
   * @param url - The GitHub remote URL
   * @returns The owner and repo
   * @throws Error if the URL cannot be parsed
   */
  private parseGitHubUrl(url: string): { owner: string; repo: string } {
    // Handle SSH URLs: git@github.com:owner/repo.git
    if (url.startsWith('git@github.com:')) {
      const match = this.extractSshGitHubComponents(url);
      if (match) {
        return match;
      }
    }

    // Handle HTTPS URLs: https://github.com/owner/repo.git
    if (url.includes('github.com')) {
      const match = this.extractHttpsGitHubComponents(url);
      if (match) {
        return match;
      }
    }

    throw new Error(`Unable to parse GitHub URL: ${url}`);
  }

  /**
   * Extract owner and repo from SSH GitHub URL
   * @param url - SSH GitHub URL
   * @returns Owner and repo or null if not matched
   */
  private extractSshGitHubComponents(url: string): { owner: string; repo: string } | null {
    const match = url.match(/git@github\.com:([^/]+)\/([^.]+)(.git)?$/);
    if (match && match[1] && match[2]) {
      return {
        owner: match[1],
        repo: match[2].replace(/\.git$/, ''),
      };
    }
    return null;
  }

  /**
   * Extract owner and repo from HTTPS GitHub URL
   * @param url - HTTPS GitHub URL
   * @returns Owner and repo or null if not matched
   */
  private extractHttpsGitHubComponents(url: string): { owner: string; repo: string } | null {
    const match = url.match(/github\.com\/([^/]+)\/([^.]+)(.git)?$/);
    if (match && match[1] && match[2]) {
      return {
        owner: match[1],
        repo: match[2].replace(/\.git$/, ''),
      };
    }
    return null;
  }

  /**
   * Parse a GitLab URL to extract the project path
   * @param url - The GitLab remote URL
   * @returns The GitLab URL and project path
   * @throws Error if the URL cannot be parsed
   */
  private parseGitLabUrl(url: string): { url: string; projectPath: string } {
    // Handle SSH URLs: git@gitlab.com:group/project.git
    if (url.startsWith('git@')) {
      const sshResult = this.extractSshGitLabComponents(url);
      if (sshResult) {
        return sshResult;
      }
    }
    // Handle HTTPS URLs: https://gitlab.com/group/project.git
    else if (url.startsWith('http')) {
      const httpsResult = this.extractHttpsGitLabComponents(url);
      if (httpsResult) {
        return httpsResult;
      }
    }

    throw new Error(`Unable to extract project path from GitLab URL: ${url}`);
  }

  /**
   * Extract GitLab components from SSH URL
   * @param url - SSH GitLab URL
   * @returns GitLab URL and project path or null if not matched
   */
  private extractSshGitLabComponents(url: string): { url: string; projectPath: string } | null {
    const match = url.match(/git@([^:]+):(.+?)(.git)?$/);
    if (match && match[1] && match[2]) {
      return {
        url: `https://${match[1]}`,
        projectPath: match[2].replace(/\.git$/, ''),
      };
    }
    return null;
  }

  /**
   * Extract GitLab components from HTTPS URL
   * @param url - HTTPS GitLab URL
   * @returns GitLab URL and project path or null if not matched
   */
  private extractHttpsGitLabComponents(url: string): { url: string; projectPath: string } | null {
    try {
      const parsedUrl = new URL(url);
      const gitlabUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;

      // Remove .git suffix if present
      const pathWithoutGit = parsedUrl.pathname.replace(/\.git$/, '');
      const projectPath = pathWithoutGit.startsWith('/') ? pathWithoutGit.slice(1) : pathWithoutGit;

      if (projectPath) {
        return { url: gitlabUrl, projectPath };
      }
    } catch (error) {
      this.logError('Error parsing GitLab HTTPS URL', error);
    }

    return null;
  }

  /**
   * Detect Git platform directly from URL (static method for URL mode)
   * @param url - The MR/PR URL to analyze
   * @returns Detected platform or null if not recognized
   */
  static detectPlatform(url: string): GitPlatform | null {
    if (!url) {
      return null;
    }

    // Check for GitLab URLs
    if (url.includes('gitlab.com') || url.includes('gitlab') || url.includes('/-/merge_requests/')) {
      return 'gitlab';
    }

    // Check for GitHub URLs
    if (url.includes('github.com') && url.includes('/pull/')) {
      return 'github';
    }

    return null;
  }

  /**
   * Log error with consistent formatting
   * @param message - Error message
   * @param error - Error object
   */
  private logError(message: string, error: unknown): void {
    const errorMessage = error instanceof Error ? error.message : String(error);
    this.logger.error(`${message}: ${errorMessage}`);
  }
}

File: python/tiw/utils/git_detector.py
"""Git platform detection utilities."""

import re
import subprocess
from typing import Dict, Optional, Any

from .logging import Logger


class GitDetector:
    """Utility class for detecting Git platform properties."""
    
    def __init__(self, logger: Logger):
        """Initialize GitDetector.
        
        Args:
            logger: Logger instance for logging messages
        """
        self.logger = logger
    
    async def detect_git_platform(self) -> Optional[Dict[str, Any]]:
        """Detect Git platform from local repository.
        
        Returns:
            Dictionary with platform information or None if not detected
        """
        # First check if we're in a Git repository
        if not self._is_git_repo():
            self.logger.warn("Not a Git repository")
            return None
        
        # Get remote URL
        remote_url = self._get_remote_url()
        if not remote_url:
            self.logger.warn("No remote URL found")
            return None
        
        # Check for GitHub repository
        github_info = self._parse_github_url(remote_url)
        if github_info:
            return {
                "platform": "github",
                **github_info
            }
        
        # Check for GitLab repository
        gitlab_info = self._parse_gitlab_url(remote_url)
        if gitlab_info:
            return {
                "platform": "gitlab",
                **gitlab_info
            }
        
        # Couldn't identify the platform
        self.logger.warn(f"Unknown Git platform URL: {remote_url}")
        return None
    
    def _is_git_repo(self) -> bool:
        """Check if the current directory is a Git repository.
        
        Returns:
            True if it's a Git repository, False otherwise
        """
        try:
            result = subprocess.run(
                ["git", "rev-parse", "--is-inside-work-tree"],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                check=False
            )
            return result.returncode == 0 and result.stdout.strip() == "true"
        except Exception as e:
            self.logger.debug(f"Error checking Git repository: {str(e)}")
            return False
    
    def _get_remote_url(self) -> Optional[str]:
        """Get the remote URL of the repository.
        
        Returns:
            Remote URL or None if not found
        """
        try:
            result = subprocess.run(
                ["git", "remote", "get-url", "origin"],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                check=False
            )
            if result.returncode == 0 and result.stdout:
                return result.stdout.strip()
            return None
        except Exception as e:
            self.logger.debug(f"Error getting remote URL: {str(e)}")
            return None
    
    def _parse_github_url(self, url: str) -> Optional[Dict[str, str]]:
        """Parse a GitHub repository URL.
        
        Args:
            url: Git remote URL
            
        Returns:
            Dictionary with GitHub repository information or None if not a GitHub URL
        """
        # HTTPS URL format: https://github.com/owner/repo.git
        https_match = re.match(r'https?://github\.com/([^/]+)/([^/\.]+)(?:\.git)?/?$', url)
        if https_match:
            return {
                "owner": https_match.group(1),
                "repo": https_match.group(2)
            }
        
        # SSH URL format: git@github.com:owner/repo.git
        ssh_match = re.match(r'git@github\.com:([^/]+)/([^/\.]+)(?:\.git)?/?$', url)
        if ssh_match:
            return {
                "owner": ssh_match.group(1),
                "repo": ssh_match.group(2)
            }
        
        return None
    
    def _parse_gitlab_url(self, url: str) -> Optional[Dict[str, str]]:
        """Parse a GitLab repository URL.
        
        Args:
            url: Git remote URL
            
        Returns:
            Dictionary with GitLab repository information or None if not a GitLab URL
        """
        # Get hostname from URL
        hostname = None
        project_path = None
        
        # HTTPS URL format: https://gitlab.com/group/project.git
        https_match = re.match(r'https?://([^/]+)/(.+?)(?:\.git)?/?$', url)
        if https_match:
            hostname = https_match.group(1)
            project_path = https_match.group(2)
        
        # SSH URL format: git@gitlab.com:group/project.git
        ssh_match = re.match(r'git@([^:]+):(.+?)(?:\.git)?/?$', url)
        if ssh_match:
            hostname = ssh_match.group(1)
            project_path = ssh_match.group(2)
        
        # Check if it's a GitLab URL
        if hostname and ('gitlab' in hostname or self._is_gitlab_instance(hostname)):
            gitlab_url = f"https://{hostname}"
            return {
                "url": gitlab_url,
                "project_path": project_path
            }
        
        return None
    
    def _is_gitlab_instance(self, hostname: str) -> bool:
        """Check if the hostname is a GitLab instance.
        
        Args:
            hostname: Hostname to check
            
        Returns:
            True if it's a GitLab instance, False otherwise
        """
        # This is a simplistic check, could be improved with API validation
        try:
            result = subprocess.run(
                ["curl", "-s", "-o", "/dev/null", "-w", "%{http_code}", f"https://{hostname}/api/v4/version"],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                check=False
            )
            return result.returncode == 0 and result.stdout.strip() in ("200", "401", "403")
        except Exception:
            return False